crates/docs/src/validation.rs

```diff
@@ -0,0 +1,70 @@
+use regex::Regex;
+use std::{
+    env, fs, io,
+    path::{Path, PathBuf},
+};
+
+pub fn extract_matches_from_file(
+    file_path: &Path,
+    re: &Regex,
+) -> Result<Vec<String>, std::io::Error> {
+    let content = fs::read_to_string(file_path)?;
+    let matches: Vec<String> = re
+        .captures_iter(&content)
+        .filter_map(|caps| caps.get(1).map(|m| m.as_str().to_string()))
+        .collect();
+
+    Ok(matches)
+}
+
+pub fn extract_matches_from_directory(
+    dir_path: &Path,
+    re: &Regex,
+    extension: Option<&str>,
+) -> io::Result<Vec<String>> {
+    let mut all_snippets = Vec::new();
+
+    for entry in fs::read_dir(dir_path)? {
+        let path = entry?.path();
+
+        if path.is_dir() {
+            all_snippets.extend(extract_matches_from_directory(&path, re, extension)?);
+        } else if extension.map_or(true, |ext| {
+            path.extension().and_then(|path_ext| path_ext.to_str()) == Some(ext)
+        }) {
+            let snippets = extract_matches_from_file(&path, re)?;
+            all_snippets.extend(snippets);
+        }
+    }
+
+    Ok(all_snippets)
+}
+
+#[must_use]
+pub fn get_parent_dir(levels_up: usize) -> PathBuf {
```

Comment:
If this is only used for tests, can we make it available only to tests? Perhaps some combination of `#[cfg(test)]` could do that?